BUG: handle introsort depth limit properly

juliantaylor opened the pull request:
each partition needs its own depth limit else it grows faster than
logarithmic ...

juliantaylor committed:
BUG: handle introsort depth limit properly

the larger partition needs its own depth limit else it grows faster than
logarithmic.
Also increase the stack size to hold up to 64 instead of 50 pointer
pairs else it might overflow on arrays larger than than 2^50 elements.

charris commented:
I've taken a closer look at this and I'm not convinced that it is correct. Note that the largest partition gets pushed on the stack so that the stack never grows rapidly, in fact it grows fastest if each partitioning exactly divides the working partition into two equal pieces. I think that what you want to check is if it is growing too slowly, which implies that the partitions on the stack are larger than desired. What am I missing?

charris commented:
Or you could push the smallest partition on the stack instead, making it grow too fast, no need for another stack in that case. If you want to duplicate the introsort pseudo code, just push the first partition on the stack along with the depth, no need for an extra stack. There are a lot of ways to do it...

charris commented:
Note that if you don't push the largest partition on the stack you will want a bigger stack. Using our own stack rather than straight recursion adds speed, so we should keep it, but it complicates introsort. Note that if you push the smallest partitions a lot of them will already be sorted one element partitions by the time you hit maximum depth.

charris commented:
I think the thing to do is ditch the new stack, decrement the depth after the partition, and put the depth on stack along with the largest partition, then reset the depth from the stack when the stack is popped. The depth popped will not always be the same as the actual stack depth, which is as it should be as it will vary with the pattern of pushes and pops.

juliantaylor commented:
it should be sufficient to only increase the depth counter on the large partition as the small partition is always less than 50% of the data and thus not degraded.

juliantaylor commented:
to clarify, the proper mirroring of the recursive intro sort would be to check inside the while loop and increment the local `cdepth` counter on each iteration. The while loop is the tail call optimized small partition of the recursive quicksort so it is bounded to log(N) stack space.
But that should not be necessary to count the small partition steps as we only want to count the maximum recursion depth which is defined by the large partition part.

pv commented:
Maybe also add benchmarks for the quicksort worst case + random input?

charris commented:
Yes, the depth counter only matters when the depth is popped off the stack which is why it should be decrementing before pushing on the stack. If the check is inside the loop insertion sort will be called before heapsort except possibly after a stack pop. And please put everything on one stack, using a separate integer stack doesn't save much room.

charris commented:
Maybe only check the depth immediately after the stack pop.

juliantaylor commented:
hm the depth check could indeed be put behind the stack pop. But it is a quite minor optimization, the check is only true in rare exceptional cases in the first place, it shouldn't make much difference if one calls insertion sort or heapsort on the small remaining data then.

charris commented:
> shouldn't make much difference if one calls insertion sort or heapsort on the small remaining data then.

Agree, it would probably never be noticed.

charris commented:
Does have the advantage that there is no goto and the while loop always runs to completion.

juliantaylor commented:
you still need the goto as you have to skip the sorting of the already heapsorted array and get to the next stack pop
I also noticed the initialization of the depth stack is unnecessary, I'll update that later
I don't like merging into one stack very much, it would then contain values interpreted as pointers and values interpreted as integer which while technically the same does not really make the code nicer.

charris commented:
OK, so in the original, just
```
            if (pi - pl < pr - pi) {
                *sptr++ = pi + 1;
                *sptr++ = pr;
                pr = pi - 1;
            }
            else {
                *sptr++ = pl;
                *sptr++ = pi - 1;
                pl = pi + 1;
            }
            *sptr++ = --depth;
```
and add
```
        if (sptr == stack) {
            break;
        }
        pr = *(--sptr);
        pl = *(--sptr);
        depth = *(--sptr)
```
Update the comment before the stack push and increase the stack size by 50, and I think that does it.

charris commented:
> it would then contain values interpreted as pointers and values 

Oh, right. I missed that, forgot we were pushing pointers instead of indices. Could have done indices originally, I guess, but the pointers were a bit faster...

charris commented:
Another option is to call heapsort rather than pushing the partition on the stack.

juliantaylor committed:
BENCH: add more sort benchmarks

benchmark for random, all equal elements, many equal elements and median
of 3 worst case

juliantaylor commented:
I have removed the initialization of the depth stack and added some new benchmarks
I did not move the heapsort bail, having a sort after the stack management code does not match the flow of the function.
Avoiding a small heapsort for a very rare case is also not worth adding a (cheap) check to all cases.

charris commented on numpy/core/src/npysort/quicksort.c.src:
```diff
@@ -107,6 +111,7 @@ quicksort_@suff@(void *start, npy_intp num, void *NOT_USED)
                 *sptr++ = pi - 1;
                 pl = pi + 1;
             }
+            *psdepth++ = cdepth + 1;
```
I think this should be `*psdepth++ = ++cdepth;`. cdepth should be increasing up the stack.

juliantaylor replied:
its `cdepth + 1`, the other variables are also changed without the postfix so this is consistent

charris replied:
I'm pretty sure it is wrong. Every push on the stack should increment, depth has nothing to do with actual stack depth here.

charris replied:
Just as every recursion in the introsort pseudo code decrements depth.

juliantaylor replied:
every stack push does increase the depth, we are only skipping the increase in the shorter partition as it is not necessary to keep track of it there

the depth is never decreased the only thing that is done is switch to a different depth in a different "recursive" calltree

charris replied:
The smaller partition never cares about depth, but the partitions on the stack do.

juliantaylor replied:
which is why it is increased and put on the stack, I don't see what you are getting at

maybe we should change it to recursive? the iterative conversion may not be worth the confusion
Or as a compromise make it an out the books recursive to iterative conversion without the small partition skips the depth counter trick?

charris replied:
Maybe I'm missing something looking at the diff. But each push should have a depth increase so that depth is monotone up the stack. Maybe you increase it somewhere else also?

charris replied:
The inplace stack does speed things up over a function call, or at least did. The function is already pretty optimized so small changes make a visible difference.

juliantaylor replied:
the confusion is probably in what each variable is.
the depth stack pointer is `psdepth` pointing to the top of the depth stack (name `depth` above)
`cdepth` is the value _on_ top the stack

     *psdepth++ = cdepth + 1

so here we increase the current depth by one, increase the stack pointer and put the new increased depth onto it.
may be rename the stack depth variable to `pdepth_stack` and the current value to `current_depth`?

charris replied:
In the pseudo code the depth is decreased (increased here) on every recursion. You can think of each pass through the loop as a recursion, so depth needs to increase on each pass.

charris replied:
The stackpointer isn't confusing, but cdepth needs to increase on every pass through the loop while nothing is getting popped, hence the `++`.

juliantaylor replied:
I'm not so sure about the performance impact of the recursion, I would not expect it to be large.
The bigger concern when implementing quicksort is the stack space usage. In order to have a limit on the stack space the compiler has to be able to do tail call optimization, having your sort crash depending on the ability of the compiler is probably worse than the quadratic performance of this case.

juliantaylor replied:
ok maybe lets backup, what do you mean with this statement in the first comment:

    *psdepth++ = ++cdepth;

this is the semantically the same as the code in this diff:

    *psdepth++ = cdepth + 1;

maybe you are looking at another broken part of the diff?

charris replied:
Not quite, it is semantically equivalent to
```
cdepth += 1;
*psdepth++ = cdepth;
``
i.e. cdepth itself is increased for the next pass through. There is no worry about actual stack depth, it is never larger than log2(n) because the largest partition gets pushed, so the loop will terminate in at most log2(n) passes, where n is the size on entry to the loop.

Thinking about this is making my head hurt ;)

juliantaylor replied:
doh right I moved this variable into a higher scope ...
it does not really matter as we don't really care about the absolute depth of the stack we just want to cut of at some point
But it is relevant to avoid the overflow of the bounded to 100 elements stack which could happen for arrays greater than 2^50 elements so I'll change it
sorry for being so dense here

juliantaylor replied:
hm the overflow can still happen as we aren't checking the depth in the small partition, so lets just increase the stack from 100 to 128 (as each level adds two pointers)

charris replied:
Sounds good.

charris replied:
Still need the `++` though.

juliantaylor replied:
yes updated the code, still got the goto though
I'll sleep about it lets see how I like it tomorrow :)

charris commented on numpy/core/src/npysort/quicksort.c.src:
```diff
@@ -69,10 +70,13 @@ quicksort_@suff@(void *start, npy_intp num, void *NOT_USED)
 
     for (;;) {
-        if (depth_limit-- < 0) {
+        if (cdepth >= depth_limit) {
```
`unlikely`? Could still move this after the pop to make the jump shorter.

juliantaylor replied:
hm yes an unlikely makes sense here

to my knowledge the length of a jump does not really matter (if both locations are in cache) and location in code has very little influence on where code ends up after compilation

charris commented on numpy/core/src/npysort/quicksort.c.src:
```diff
@@ -198,6 +208,7 @@ stack_pop:
         }
         pr = *(--sptr);
         pl = *(--sptr);
+        cdepth = *(--psdepth);
```
For instance
```
                while (sptr != stack) {
                     pr = *(--sptr);
                     pl = *(--sptr);
                     cdepth = *(--psdepth) 
                     if (cdepth < depth_limit) {
                         break;
                     }
                     aheapsort_@suff@(vv, pl, pr - pl + 1, NULL);
                 }
                 if (sptr == stack) {
                     break;
                 }
```
which gets rid of the goto.

juliantaylor replied:
but this switches everything to heapsort on the first bad partition, a bad partition does not mean the full array needs to be heapsorted only the current part

charris replied:
No it doesn't, if `depth` is properly incremented it loops at most twice.

charris replied:
Note that the logic is basically what you have now, just no `goto`.

juliantaylor replied:
hm right I misread the break condition.
I still don't like it as much. The goto variant makes the code structure more similar to the recursive code which is much easier to understand, you just have to interpret the goto stack pop as the equivalent of the return statement in the recursive one.
But if you prefer it like that I can change it

charris replied:
Well, it has the advantage the need to increment depth on every push is made clear. I think it makes the logic a bit clearer because everything takes place in a few lines. Note that you do need the `++depth` so that the depth is strictly increasing up the stack so only the top partition can be heapsorted on any pass through.

juliantaylor replied:
I feel the opposite, I have to think about your code is doing while the goto is comes natural to me, but that might be from my close to the machine code perspective where a function return is a jump to the stack pop.

charris replied:
Yeah, the goto goes back instead of up. I don't think either is crystalline clear, having the stack the way we do breaks the obviousness of the doubly recursive pseudo code. Even pushing the largest partition is a bit subtle.

juliantaylor commented:
I'll switch it to decreasing the depth down from the limit, I initially increased it because I misguidedly though I had to initialize the depth stack (so memset(0) vs a for loop)
Then the pseudo code of the recursive implementation matches the iterative variant so I'll add that as a comment which should help clearing up how this works.

charris commented:
I'm thinking that instead of `npy_get_msb`, make the stack of `size_t` type, init with double the count and shift right on each pass instead of decrementing. That solves the overflow problem and is also more flexible.

charris commented on numpy/core/src/npysort/quicksort.c.src:
```diff
@@ -69,10 +90,12 @@ quicksort_@suff@(void *start, npy_intp num, void *NOT_USED)
-    npy_intp depth_limit = npy_get_msb(num) * 2;
+    int depth[PYA_QS_STACK];
+    int * psdepth = depth;
+    int cdepth = NPY_SORT_MIN(PYA_QS_STACK, npy_get_msb(num) * 2);
```
I don't think `* 2` is needed here, just adjust the cutoff point. Then you don't need the `NPY_SORT_MIN` either, especially as it is now `64`. I suppose you could also use `NPY_BITSOF_INTP`

charris replied:
NVM. However, the stack can't fill up, so just
```
int cdepth = npy_get_msb(num) * 2;
```

juliantaylor replied:
the factor 2 is just some random increase (which should be documented) to give the algorithm some slack on not ideal data. The only thing that needs to be done is avoid that the depth is almost linear in number of elements so as long as the limit is constant * log2(n) we should have our worst case.
but the min can now be removed

charris replied:
Yep, cdepth can be anything that serves as a reasonable cutoff. It is independent of the actual stack depth as there may be jumps in it's value greater than one along the stack: partition gets popped, gets replaced by partition with cdepth decremented. Continuous bad partitioning actually leads to the smallest stack.

charris commented on numpy/core/src/npysort/quicksort.c.src:
```diff
@@ -69,10 +90,12 @@ quicksort_@suff@(void *start, npy_intp num, void *NOT_USED)
 
     for (;;) {
-        if (depth_limit-- < 0) {
+        if (NPY_UNLIKELY(cdepth < 0)) {
```
The pseudocode uses ` <= 0`, any reason for the change? Note that this would work fine if you didn't multiply by two above.

charris replied:
Note that the loop terminates early due to insertion sort, but I don't think that is a problem, it's the big partitions we worry about and they should be handled correctly in that situation as when they are popped they will probably go through the loop.

charris replied:
Doesn't matter much, actually, adjustable.

juliantaylor replied:
as the stack now cannot overflow anymore it doesn't matter if we abort one earlier or later, I'll adjust the pseudocode to match

charris replied:
Yeah, I don't think it matters much given the factor of 2 already in there.

charris commented:
Looks generally good, just a few comments for consideration.

charris commented on numpy/core/src/npysort/quicksort.c.src:
```diff
@@ -33,10 +53,11 @@
 
 #define NOT_USED NPY_UNUSED(unused)
-#define PYA_QS_STACK 100
+#define PYA_QS_STACK 128
```
Maybe 
```
#define PYA_QS_STACK 2*NPY_BITSOF_INTP;
```
Or just use NPY_BITSOF_INTP directly below.

juliantaylor commented:
hm shifting the count is a neat way of writing it, but I think it makes it less obvious what we are doing. Performance wise it shouldn't matter, the msb is cheap (and could be further optimized using `__builtin_clz`)

charris commented:
> hm shifting the count is a neat way of writing it, 

On second thought I decided it was a bad idea ;) It can't get the factor of two increase in bits.

charris commented on numpy/core/src/npysort/quicksort.c.src:
```diff
@@ -69,10 +94,12 @@ quicksort_@suff@(void *start, npy_intp num, void *NOT_USED)
     @type@ **sptr = stack;
     @type@ *pm, *pi, *pj, *pk;
-    npy_intp depth_limit = npy_get_msb(num) * 2;
+    int depth[PYA_QS_STACK];
```
Well, this is large by a factor of 2, but it certainly doesn't hurt.

charris commented:
Thanks Julian.
